File: Solutions/17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.rb
def letter_combinations(digits)
    return [] if digits.empty?
    
    phone_map = {
        '2' => 'abc',
        '3' => 'def',
        '4' => 'ghi',
        '5' => 'jkl',
        '6' => 'mno',
        '7' => 'pqrs',
        '8' => 'tuv',
        '9' => 'wxyz'
    }
    
    @output = []
    
    def backtrack(combination, next_digits, phone_map)
        if next_digits.empty?
            @output << combination
        else
            phone_map[next_digits[0]].each_char do |letter|
                backtrack(combination + letter, next_digits[1..-1], phone_map)
            end
        end
    end
    
    backtrack("", digits, phone_map)
    @output
end

File: Solutions/17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cs
public class Solution {
    private Dictionary<char, string> phoneMap = new Dictionary<char, string>() {
        {'2', "abc"},
        {'3', "def"},
        {'4', "ghi"},
        {'5', "jkl"},
        {'6', "mno"},
        {'7', "pqrs"},
        {'8', "tuv"},
        {'9', "wxyz"}
    };
    
    private List<string> output = new List<string>();
    
    public IList<string> LetterCombinations(string digits) {
        if (string.IsNullOrEmpty(digits)) {
            return new List<string>();
        }
        
        Backtrack("", digits);
        return output;
    }
    
    private void Backtrack(string combination, string nextDigits) {
        if (nextDigits.Length == 0) {
            output.Add(combination);
        }
        else {
            foreach (char letter in phoneMap[nextDigits[0]]) {
                Backtrack(combination + letter, nextDigits.Substring(1));
            }
        }
    }
}